cpp/src/datetime/datetime_ops.cu

```diff
@@ -263,29 +268,66 @@ std::unique_ptr<column> add_calendrical_months(column_view const& timestamp_colu
-  if (size == 0) return make_empty_column(output_col_type);
-
-  auto output_col_mask =
-    cudf::detail::bitmask_and(table_view({timestamp_column, months_column}), stream, mr);
-  auto output = make_fixed_width_column(
-    output_col_type, size, std::move(output_col_mask), cudf::UNKNOWN_NULL_COUNT, stream, mr);
-
-  auto launch = add_calendrical_months_functor{
-    timestamp_column, months_column, static_cast<mutable_column_view>(*output)};
-
-  type_dispatcher(timestamp_column.type(), launch, stream);
 
+  auto output = months_column.type() == data_type{type_id::INT16}
+                  ? type_dispatcher(timestamp_column.type(),
+                                    add_calendrical_months_functor{},
+                                    timestamp_column,
+                                    months_column.begin<int16_t>(),
+                                    stream,
+                                    mr)
+                  : type_dispatcher(timestamp_column.type(),
+                                    add_calendrical_months_functor{},
+                                    timestamp_column,
+                                    months_column.begin<int32_t>(),
+                                    stream,
+                                    mr);
+
+  auto output_null_mask =
+    cudf::detail::bitmask_and(table_view{{timestamp_column, months_column}}, stream, mr);
+  output->set_null_mask(std::move(output_null_mask));
   return output;
 }
 
+std::unique_ptr<column> add_calendrical_months(column_view const& timestamp_column,
+                                               scalar const& months,
+                                               rmm::cuda_stream_view stream,
+                                               rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(is_timestamp(timestamp_column.type()), "Column type should be timestamp");
+  CUDF_EXPECTS(months.type().id() == type_id::INT16 or months.type().id() == type_id::INT32,
+               "Months type should be INT16 or INT32");
+
+  if (months.is_valid(stream)) {
+    auto output =
+      months.type().id() == type_id::INT16
+        ? type_dispatcher(timestamp_column.type(),
+                          add_calendrical_months_functor{},
+                          timestamp_column,
+                          thrust::make_constant_iterator(
+                            static_cast<numeric_scalar<int16_t> const&>(months).value(stream)),
+                          stream,
+                          mr)
+        : type_dispatcher(timestamp_column.type(),
+                          add_calendrical_months_functor{},
+                          timestamp_column,
+                          thrust::make_constant_iterator(
+                            static_cast<numeric_scalar<int32_t> const&>(months).value(stream)),
+                          stream,
+                          mr);
+    output->set_null_mask(cudf::detail::copy_bitmask(timestamp_column, stream, mr));
```

Comment:
This should be done inside the `add_calendrical_months_functor`.

Reply:
I don't think it can. `add_calendrical_months_functor` is shared by both the column and scalar API. The column API requires the output bitmask to be a logical and between the input and `months`. And since `add_calendrical_months_functor` only accepts iterators, it has no bitmask information about the months column.